### PR TITLE
Implement feeder, transaction gateway clients and transaction table

### PR DESCRIPTION
**Pull request by torztomasz**
Resolves L2B-1441, L2B-1442

### CODE REVIEW

File: packages/backend/src/peripherals/starkware/FeederGatewayClient.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    private readonly options: GatewayConfig,
+    private readonly fetchClient: FetchClient
+  ) {
+    super(options.auth)
+  }
+
+  async getPerpetualBatchInfo(batchId: number) {
+    const data = await this.getBatchInfo(batchId)
+    const parsed = PerpetualBatchInfoResponse.parse(data)
+    return parsed // TODO: toPerpetualTransactionBatchResponse(parsed)
```
Comment (adamiak) [inline packages/backend/src/peripherals/starkware/FeederGatewayClient.ts:17]: Is this TODO still pending?

File: packages/backend/src/peripherals/starkware/TransactionGatewayClient.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+export class TransactionGatewayClient extends GatewayClient {
+  constructor(
+    private readonly options: TransactionGatewayConfig,
+    private readonly fetchClient: FetchClient
+  ) {
+    super(options.auth)
+  }
+
+  async getPerpetualTransactions(startApexId: number, expectCount: number) {
```
Comment (adamiak) [inline packages/backend/src/peripherals/starkware/TransactionGatewayClient.ts:15]: rename apexId to thirdPartyId in this whole file.

**Review by adamiak — CHANGES_REQUESTED** [review]

File: packages/backend/src/Application.ts [hunk 3]
```diff
@@ -224,7 +226,8 @@ export class Application {
 
     if (config.starkex.dataAvailabilityMode === 'validium') {
       const availabilityGatewayClient = new AvailabilityGatewayClient(
-        config.starkex.availabilityGateway
+        config.starkex.availabilityGateway,
+        fetchClient
```
[the remark this answers is not in the document]
Comment (torztomasz) [reply packages/backend/src/Application.ts:230]: Well, FetchClient needs a logger and it is easier to mock FetchClients for tests if it is done this way

File: packages/backend/src/peripherals/database/TransactionRepository.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  decodeTransactionData,
+  encodeTransactionData,
+  TransactionData,
+} from './Transaction'
+
+interface Record<T extends TransactionData['type'] = TransactionData['type']> {
+  id: number
+  thirdPartyId: number
+  transactionId: number
+  starkKeyA: StarkKey | undefined
```
[the remark this answers is not in the document]
Comment (torztomasz) [reply packages/backend/src/peripherals/database/TransactionRepository.ts:17]: well using optional parameter would suggest that sometimes this column is not present or smth, using `| undefined` tells developer that it is always present but sometimes without defined value

File: packages/backend/src/peripherals/database/migrations/035_transactions.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+import { Knex } from 'knex'
+
+const TRANSACTIONS_TABLE_NAME = 'transactions'
+
+export async function up(knex: Knex) {
+  await knex.schema.createTable(TRANSACTIONS_TABLE_NAME, (table) => {
+    table.increments('id').primary()
+    table.integer('third_party_id').unique().notNullable().index()
+    table.integer('transaction_id').notNullable().index()
```
[the remark this answers is not in the document]
Comment (torztomasz) [reply packages/backend/src/peripherals/database/migrations/035_transactions.ts:22]: Actually, there are no timestamps for all the transactions which is very weird. Maybe we can ask Jackie to add the timestamp

File: packages/backend/src/peripherals/starkware/FeederGatewayClient.ts (hunk 1 above)
Comment (torztomasz) [reply packages/backend/src/peripherals/starkware/FeederGatewayClient.ts:17]: Well, it should be implemented while working on comparing Feeder response and Transactions in database.

File: packages/backend/src/peripherals/starkware/schema.ts [hunk 6]
```diff
@@ -3,16 +3,20 @@ import * as z from 'zod'
 const UnsignedIntAsString = z.string().regex(/^([1-9]\d*|0)$/)
 const SignedIntAsString = z.string().regex(/^(-?[1-9]\d*|0)$/)
 const PedersenHash = z.string().regex(/^0[a-f\d]{63}$/)
-const PedersenHash0x = z.string().regex(/^0x[a-f\d]{0,63}$/)
+const PedersenHash0x = z.string().regex(/^0x[a-f\d]{0,64}$/)
+const StarkKey = z.string().regex(/^0x0[a-f\d]{63}$/)
+const Bytes32 = z.string().regex(/^[a-f\d]{64}$/)
 const AssetHash0x = z.string().regex(/^0x[a-f\d]{0,63}$/)
 const AssetId = z.string().regex(/^0x[a-f\d]{30}$/)
+const EthereumAddress = z.string().regex(/^0x[a-fA-F0-9]{40}$/)
```
[the remark this answers is not in the document]
Comment (torztomasz) [reply packages/backend/src/peripherals/starkware/schema.ts]: Leave {40}

File: packages/backend/src/peripherals/starkware/schema.ts [hunk 7]
```diff
@@ -3,24 +3,29 @@ import * as z from 'zod'
 const UnsignedIntAsString = z.string().regex(/^([1-9]\d*|0)$/)
 const SignedIntAsString = z.string().regex(/^(-?[1-9]\d*|0)$/)
 const PedersenHash = z.string().regex(/^0[a-f\d]{63}$/)
-const PedersenHash0x = z.string().regex(/^0x[a-f\d]{0,63}$/)
+const Hash256_0x = z.string().regex(/^0x[a-f\d]{1,64}$/)
+const Hash256 = z.string().regex(/^[a-f\d]{1,64}$/)
+const StarkKey0x = z.string().regex(/^0x[a-f\d]{1,64}$/)
+const StarkKey0x0 = z.string().regex(/^0x0[a-f\d]{63}$/)
```
Comment (adamiak) [inline packages/backend/src/peripherals/starkware/schema.ts:9]: Why do we need to regexes for StarkKey? AFAIK every StarkKey starts with 0x0

**Review by adamiak — APPROVED** [review]

**Pull request merged**